modules/c++/mt/include/mt/WorkSharingBalancedRunnable1D.h

```diff
@@ -0,0 +1,295 @@
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; If not,
+ * see <http://www.gnu.org/licenses/>.
+ *
+ */
+#ifndef __MT_WORK_SHARING_BALANCED_RUNNABLE_1D_H__
+#define __MT_WORK_SHARING_BALANCED_RUNNABLE_1D_H__
+
+#include <vector>
+#include <sstream>
+
+#include <sys/Conf.h>
+#include <sys/Runnable.h>
+#include <except/Exception.h>
+#include <mt/ThreadPlanner.h>
+#include <mt/ThreadGroup.h>
+#include <types/Range.h>
+
+namespace mt
+{
+typedef std::vector<mem::SharedPtr<sys::AtomicCounter> > SharedAtomicCounterVec;
+
+/*!
+ *  This runnable takes both a reference to the associated thread's atomic
+ *  counter as well as atomic counter and range information for all other
+ *  threads in the thread pool.
+ *
+ *  Each runnable will operate on a contiguous range of elements
+ *  ([startElement, startElement + numElements]). Once all work has been
+ *  completed in its range, the atomic counters and terminating index for
+ *  the other threads will be used to grab additional work rather than let
+ *  the thread die.
+ *
+ *  This runnable is useful in cases where work needs to be
+ *  done across a range of elements, but when dividing these elements
+ *  across threads leads to balancing issues i.e certain threads
+ *  terminate earlier than other threads. By giving each thread
+ *  its own range and atomic counter rather than one shared atomic counter
+ *  used to grab elements from a global range, we get better locality of
+ *  reference and in practice better caching.
+ *
+ *  \param range Range of elements for this runnable to work on
+ *
+ *  \param[in,out] atomicCounter Atomic counter this thread will initially use to
+ *  fetch elements within its range to process
+ *
+ *  \param threadPoolCounters Atomic counters of other threads - if
+ *  this thread finishes processing the elements within its range, these
+ *  counters will be used to grab additional work
+ *
+ *  \param threadPoolEndElements Terminating indices (exclusive) for each
+ *  thread
+ *
+ */
+template <typename OpT>
+class WorkSharingBalancedRunnable1D : public sys::Runnable
+{
+public:
+        WorkSharingBalancedRunnable1D(
+                const types::Range& range,
+                sys::AtomicCounter& counter,
+                const SharedAtomicCounterVec& threadCounters,
+                const std::vector<size_t>& threadPoolEndElements,
+                const OpT& op) :
+        mStartElement(range.mStartElement),
+        mEndElement(mStartElement + range.mNumElements),
+        mCounter(counter),
+        mThreadPoolCounters(threadCounters),
+        mThreadPoolEndElements(threadPoolEndElements),
+        mOp(op)
+    {
+    }
+
+    virtual void run()
+    {
+        // Operate over this thread's range
+        while (true)
+        {
+            const size_t element = mCounter.getThenIncrement();
```

Comment:
Not that this logic is super complicated, but we might as well not repeat it... the whole `while` block is repeated here and below.  Could have a private method that takes an atomic counter and an end element and does this while block, then call it in both spots